core/server/api/v0.1/decorators/urls.js

```diff
@@ -35,53 +35,57 @@ const urlsForPost = (post, options) => {
             // @NOTE: this block also decorates primary_tag/primary_author objects as they
             // are being passed by reference in tags/authors. Might be refactored into more explicit call
             // in the future, but is good enough for current use-case
-            if (relation === 'tags' && post.tags) {
-                post.tags = post.tags.map(tag => urlsForTag(tag, options));
+            if (relation === 'tags' && attrs.tags) {
+                attrs.tags = attrs.tags.map(tag => urlsForTag(tag.id, tag, options));
             }
 
-            if (relation === 'author' && post.author) {
-                post.author = urlsForUser(post.author, options);
+            if (relation === 'author' && attrs.author) {
+                attrs.author = urlsForUser(attrs.author.id, attrs.author, options);
             }
 
-            if (relation === 'authors' && post.authors) {
-                post.authors = post.authors.map(author => urlsForUser(author, options));
+            if (relation === 'authors' && attrs.authors) {
+                attrs.authors = attrs.authors.map(author => urlsForUser(author.id, author, options));
             }
         });
     }
 
-    return post;
+    return attrs;
 };
 
-const urlsForUser = (user, options) => {
+const urlsForUser = (id, attrs, options) => {
     if (options && options.context && options.context.public && options.absolute_urls) {
-        user.url = urlFor({
-            relativeUrl: urlService.getUrlByResourceId(user.id)
-        }, true);
+        if (id) {
+            attrs.url = urlFor({
+                relativeUrl: urlService.getUrlByResourceId(id)
+            }, true);
+        }
 
-        if (user.profile_image) {
-            user.profile_image = urlFor('image', {image: user.profile_image}, true);
+        if (attrs.profile_image) {
+            attrs.profile_image = urlFor('image', {image: attrs.profile_image}, true);
         }
 
-        if (user.cover_image) {
-            user.cover_image = urlFor('image', {image: user.cover_image}, true);
+        if (attrs.cover_image) {
+            attrs.cover_image = urlFor('image', {image: attrs.cover_image}, true);
         }
     }
 
-    return user;
+    return attrs;
 };
 
-const urlsForTag = (tag, options) => {
+const urlsForTag = (id, attrs, options) => {
     if (options && options.context && options.context.public && options.absolute_urls) {
-        tag.url = urlFor({
-            relativeUrl: urlService.getUrlByResourceId(tag.id)
-        }, true);
+        if (id) {
```

Comment:
Why checking for "if id"?

Reply:
Otherwise, we'd get a `/404/` in the url just like in the original bug. The only place the id might not be passed is in the call from here: https://github.com/TryGhost/Ghost/pull/9969/files#diff-bc8a5cba05e22f60f0bd0a02d10203ebR39

Reply:
Ups, hm then i think we have to organise the code differently to ensure the resource.id is always accessible.

Reply:
Agree. This is exactly why left this comment - https://github.com/TryGhost/Ghost/pull/9969#discussion_r223634149. 

Passing even more id's here doesn't look nice and would add quite a lot of glue code in the controller level if we go that way. On the other hand, there's no other way of passing these values in if we call `toJSON` before calling `urlsForPost`. 

Another approach could be actually calling `urlsForPost` with a raw model (where id is always available) and returning an object with url's that would be merged with the result of `toJSON`:
```javascript
posts: data.map((model) => {
    const urlAttrs = urlsForPost(model, options);
    return _.merge(model.toJSON(), urlAttrs);
})
```
Haven't checked how the merge would work in this case on nested tags/authors objects.

Reply:
Yeah i think that makes sense 👍

Reply:
Discussed: Figure how to current behaviour is when requesting `fields=url&include=tags` - will v0.1 only return `tags.url`?

Discussed: We will def fix that on Monday.

Discussed: We also have to update v2, because it will just copy the v0.1 decoration logic 👍

Reply:
Did a little investigation into how options were passed in older versions and how URLs were generated. Summary:
- it wasn't possible to fetch child elements when `fields` option was provided
- because of the above, `url` attribute field filter only worked on the parent level

In current implementation:
- the bug wouldn't actually effect nested properties because we don't seem to apply the filter to it (all of the properties are returned)